## 1. What is a neural network? What are the general steps required to build a neural network? 

A neural network is a series of algorithms that can effectively recognize underlying relationships in a set of data through a process that mimics the way the human brain operates. Neural network is a powerful modeling approach that accounts for interactions between data really well.

<b>The general steps required to build a neural network are:</b>

* put together an input layer and define number of input nodes
* a hidden layer or layers and define the number of nodes in these hidden layers
* define the number of ourput nodes; number of output classes you want the neural network to process
* define an activation function for each layer, for example, relu or linear

## 2. Generally, how do you check the performance of a neural network? Why?

To check the performance , we can divide the training set in a real training set and a validation set using one of these methods:

* Cross-validation
* stratified holdout
* 0.632 bootstrap

And then we can measure the performance of the ANN with one of these metrics: TP rate, FP rate, F-measure, accuracy
precision and recall.

## 3. Create a neural network using keras to predict the outcome of either of these datasets: 
Cardiac Arrhythmia: https://archive.ics.uci.edu/ml/datasets/Arrhythmia 
Abalone age: https://archive.ics.uci.edu/ml/datasets/Abalone

In [76]:
import numpy as np
import pandas as pd
import pandas as pd
from keras.layers import Dense
from keras.models import Sequential
from keras.utils.np_utils import to_categorical
from sklearn.preprocessing import StandardScaler

In [73]:
#red the dataset file
abalone_df = pd.read_csv("abalone_edited.csv")

In [74]:
abalone_df.head()

Sex  lengths  Diameter  height  Whole weight  Shucked weight  \
0   M    0.455     0.365   0.095        0.5140          0.2245   
1   M    0.350     0.265   0.090        0.2255          0.0995   
2   F    0.530     0.420   0.135        0.6770          0.2565   
3   M    0.440     0.365   0.125        0.5160          0.2155   
4   I    0.330     0.255   0.080        0.2050          0.0895   

   Viscera weight  Shell weight  Rings  
0          0.1010         0.150     15  
1          0.0485         0.070      7  
2          0.1415         0.210      9  
3          0.1140         0.155     10  
4          0.0395         0.055      7

In [77]:
#checking data types
abalone_df.dtypes

Sex                object
lengths           float64
Diameter          float64
height            float64
Whole weight      float64
Shucked weight    float64
Viscera weight    float64
Shell weight      float64
Rings               int64
dtype: object

In [78]:
#dropping "sex" cilumn
new = abalone_df.drop("Sex", axis=1)

In [79]:
#scaling data
scaler = StandardScaler()
df_scaled = pd.DataFrame(scaler.fit_transform(new), columns = new.columns)

In [80]:
df_scaled.head()

lengths  Diameter    height  Whole weight  Shucked weight  Viscera weight  \
0 -0.574558 -0.432149 -1.064424     -0.641898       -0.607685       -0.726212   
1 -1.448986 -1.439929 -1.183978     -1.230277       -1.170910       -1.205221   
2  0.050033  0.122130 -0.107991     -0.309469       -0.463500       -0.356690   
3 -0.699476 -0.432149 -0.347099     -0.637819       -0.648238       -0.607600   
4 -1.615544 -1.540707 -1.423087     -1.272086       -1.215968       -1.287337   

   Shell weight     Rings  
0     -0.638217  1.571544  
1     -1.212987 -0.910013  
2     -0.207139 -0.289624  
3     -0.602294  0.020571  
4     -1.320757 -0.910013

In [181]:
X = df_scaled.drop('Rings', axis=1)
y = df_scaled['Rings']

In [182]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=42)

In [183]:
#building a model
model = Sequential()
model.add(Dense(10, activation='relu'))
model.add(Dense(1, activation = 'linear'))

In [184]:
model.compile(optimizer='adam', loss='mean_squared_error', metrics=['mae', 'RootMeanSquaredError'])

In [185]:
model.fit(X_train, y_train, batch_size=20, epochs = 50)

Epoch 1/50
168/168 [==============================] - 1s 1ms/step - loss: 1.1237 - mae: 0.8263 - root_mean_squared_error: 1.0562
Epoch 2/50
168/168 [==============================] - 0s 1ms/step - loss: 0.6674 - mae: 0.5830 - root_mean_squared_error: 0.8161
Epoch 3/50
168/168 [==============================] - 0s 1ms/step - loss: 0.6344 - mae: 0.5777 - root_mean_squared_error: 0.7959
Epoch 4/50
168/168 [==============================] - 0s 1ms/step - loss: 0.5723 - mae: 0.5543 - root_mean_squared_error: 0.7559
Epoch 5/50
168/168 [==============================] - 0s 1ms/step - loss: 0.5459 - mae: 0.5467 - root_mean_squared_error: 0.7384
Epoch 6/50
168/168 [==============================] - 0s 1ms/step - loss: 0.5458 - mae: 0.5334 - root_mean_squared_error: 0.7386
Epoch 7/50
168/168 [==============================] - 0s 2ms/step - loss: 0.5429 - mae: 0.5353 - root_mean_squared_error: 0.7363
Epoch 8/50
168/168 [==============================] - 0s 2ms/step - loss: 0.5219 - mae: 0.5171 - 

In [194]:
from sklearn.metrics import classification_report
y_pred = model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           2       0.00      0.00      0.00         1
           3       0.00      0.00      0.00         6
           4       0.47      0.28      0.35        29
           5       0.34      0.37      0.36        54
           6       0.37      0.27      0.31       142
           7       0.32      0.47      0.38       179
           8       0.33      0.31      0.32       293
           9       0.26      0.41      0.32       339
          10       0.23      0.31      0.27       315
          11       0.26      0.34      0.29       250
          12       0.00      0.00      0.00       134
          13       0.06      0.02      0.03        99
          14       0.16      0.06      0.08        54
          15       0.00      0.00      0.00        56
          16       0.10      0.06      0.07        34
          17       0.00      0.00      0.00        33
          18       0.00      0.00      0.00        22
          19       0.00    

/Users/olesiamukha/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/olesiamukha/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/olesiamukha/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1248: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## 4. Write another algorithm to predict the same result as the previous question using either KNN or logistic regression.

In [104]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

In [202]:
#logostic regression
X = df_scaled.drop('Rings', axis=1) #scaling X
y = new['Rings']

#split data into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=50)


model = LogisticRegression(random_state=42)
model.fit(X_train, y_train)
model.score(X_test, y_test)

/Users/olesiamukha/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.27190043082814747

In [203]:
from sklearn.metrics import mean_squared_error
y_pred = model.predict(X_test)
mean_squared_error(y_test, y_pred)

6.749162278602202

## 5. Create a neural network using pytorch to predict the same result as question 3

In [196]:
import torch
import torch.nn as nn
import torch.nn.functional as F #this has activation functions

X = df_scaled.drop('Rings', axis=1) #scaling X
y = df_scaled['Rings']

#split data into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=50)


# Creating tensors
X_train = torch.LongTensor(X_train.to_numpy())
X_test = torch.LongTensor(X_test.to_numpy())

y_train = torch.FloatTensor(y_train.to_numpy())
y_test = torch.FloatTensor(y_test.to_numpy())

print(X_train)

tensor([[ 0,  0,  0,  ...,  0,  0,  0],
        [ 0,  0,  0,  ...,  0,  0,  0],
        [ 0,  0,  0,  ...,  0,  0,  0],
        ...,
        [-1, -1, -1,  ..., -1, -1, -1],
        [ 0,  0,  0,  ...,  0,  0,  0],
        [ 0,  0,  0,  ...,  0,  1,  0]])


In [197]:
class ANN_Model(nn.Module):
    def __init__(self, input_features=7, hidden1=20, hidden2=20, out_features =2):
        super().__init__()
        self.layer_1_connection = nn.Linear(input_features, hidden1)
        self.layer_2_connection = nn.Linear(hidden1, hidden2)
        self.out = nn.Linear(hidden2, out_features)
    
    def forward(self, x):
        #apply activation functions
        x = F.relu(self.layer_1_connection(x))
        x = F.relu(self.layer_2_connection(x))
        x = self.out(x)
        return x

In [198]:
torch.manual_seed(42)

#instantiate the model
model = ANN_Model()

In [199]:
# loss function
loss_function = nn.CrossEntropyLoss()

#optimizer
optimizer = torch.optim.Adam(model.parameters(), lr = 0.01)

In [200]:
#run model through multiple epochs/iterations
final_loss = []
n_epochs = 100
for epoch in range(n_epochs):
    y_pred = model.forward(X_train)
    loss = loss_function(y_pred, y_train)
    final_loss.append(loss)
    
    if epoch % 10 == 1:
        print(f'Epoch number: {epoch} with loss: {loss.item()}')
    
    optimizer.zero_grad() #zero the gradient before running backwards propagation
    loss.backward() #for backward propagation 
    optimizer.step() #performs one optimization step each epoch   

RuntimeError: expected scalar type Float but found Long

## 6. Compare the performance of the neural networks to the other model you created. Which performed better? Why do you think that is?

To evaluate the performance of the models I mainly looked at Root mean square error. Based on the abalone dataset, neural networks model performed best having **root mean square error of 0.6280.**
Logistic regression accuracy turned out to be very low for this dataset - 0.27 and root mean square error very high - 6.74.

I think neural network model performed better because it is known to be better in identifying the relationships and interactions between the model's features.